tools/infrasys/src/main.rs

```diff
@@ -0,0 +1,872 @@
+            .as_ref()
+            .context(error::MissingConfig {
+                missing: format!("vpc_endpoint_id for '{}' s3 config", s3_stack_name),
+            })?;
+        let prefix = s3_info.s3_prefix.as_ref().context(error::MissingConfig {
+            missing: format!("s3_prefix for '{}' s3 config", s3_stack_name),
+        })?;
+        let signing_keys = repo_config
+            .signing_keys
+            .as_mut()
+            .context(error::MissingConfig {
+                missing: format!("signing_keys for '{}' repo config", repo_name),
+            })?;
+        let root_keys = repo_config
+            .root_keys
+            .as_mut()
+            .context(error::MissingConfig {
+                missing: format!("root_keys for '{}' repo config", repo_name),
+            })?;
+
+        // Step 1: Create S3 Bucket
+        info!("Creating S3 bucket...");
+        let (s3_stack_arn, bucket_name, bucket_url) =
+            create_s3_bucket(s3_region, s3_stack_name).await?;
+        s3_info.stack_arn = Some(s3_stack_arn);
+        s3_info.bucket_name = Some(bucket_name.clone());
+
+        // Step 2: Add Bucket Policy
+        // TODO: test adding bucket policy without the /*
+        add_bucket_policy(s3_region, &bucket_name, prefix, vpcid).await?;
+
+        // Step 3: Create root + publication keys
+        info!("Creating KMS Keys...");
+        create_keys(signing_keys).await?; 
+        create_keys(root_keys).await?;
+
+        // Step 4: Create and sign root.json
+        info!("Creating and signing root.json...");
+        info!("Below are TUF created key_ids as we add keys to root.json:"); // QUESTION: is this accurate?
+        create_root(&root_role_path)?;
+        // Sign root (with both roles)
+        sign_root(
+            signing_keys,
+            &KeyRole::Publication,
+            repo_config
+                .pub_key_threshold
+                .as_ref()
+                .context(error::MissingConfig {
+                    missing: format!("pub_key_threshold for '{}' repo config", repo_name),
+                })?,
+            &root_role_path.display().to_string(),
+        )?;
+        sign_root(
+            root_keys,
+            &KeyRole::Root,
+            repo_config
+                .root_key_threshold
+                .as_ref()
+                .context(error::MissingConfig {
+                    missing: format!("root_key_threshold for '{}' repo config", repo_name),
+                })?,
+            &root_role_path.display().to_string(),
+        )?;
+
+        // Step 5: Upload root.json
+        info!("Uploading root.json to S3 bucket...");
+        upload_file(s3_region, &bucket_name, prefix, root_role_path).await?;
+
+        // Step 6: Update output paramters if not already set
+        if repo_config.metadata_base_url.is_none() {
+            repo_config.metadata_base_url = Some(
+                Url::parse(format!("{}{}/metadata/", &bucket_url, prefix).as_str())
```

Comment:
I recommend adding some checks to make sure the base forward slash isn't ever omitted from the url for peace of mind.
Something like this https://github.com/bottlerocket-os/bottlerocket/blob/b9c8747e7abf4f5dd02e217571ed8d26f3f3db72/tools/pubsys/src/repo.rs#L305-L311

Same below for `targets_url` and `root_role_url`